components/dash-core-components/src/components/Tooltip.react.js

```diff
@@ -7,20 +7,167 @@ import _JSXStyle from 'styled-jsx/style'; // eslint-disable-line no-unused-vars
  * A tooltip with an absolute position.
  */
 const Tooltip = props => {
-    const {bbox, border_color, background_color, id, loading_state} = props;
+    const {
+        bbox,
+        border_color,
+        background_color,
+        id,
+        loading_state,
+        auto_direction_container: autoDirectionContainer,
+    } = props;
     const is_loading = loading_state?.is_loading;
     const show = props.show && bbox;
+    const tooltipElement = useRef();
+
+    const initialDirection =
+        props.direction === 'auto' ? 'right' : props.direction;
+    const [realDirection, setRealDirection] = useState(initialDirection);
+
+    const getRealHiddenPart = value => {
+        return value > 0 ? value : 0;
+    };
+
+    const calculateHiddenPart = (parentRect, tooltipRect) => {
+        const hiddenDirections = [
+            {
+                direction: 'left',
+                value: getRealHiddenPart(parentRect.left - tooltipRect.left),
+            },
+            {
+                direction: 'right',
+                value: getRealHiddenPart(tooltipRect.right - parentRect.right),
+            },
+            {
+                direction: 'top',
+                value: getRealHiddenPart(parentRect.top - tooltipRect.top),
+            },
+            {
+                direction: 'bottom',
+                value: getRealHiddenPart(
+                    tooltipRect.bottom - parentRect.bottom
+                ),
+            },
+        ];
+
+        return hiddenDirections.reduce((sum, hidden) => sum + hidden.value, 0);
```

Comment:
This one doesn't use `direction`, right? So could just be an array of 4 numbers, then
```js
return sum(hiddenDirections); // import sum from 'ramda'
```

Reply:
also we might want to normalize by the tooltip size in each direction, ie instead of
```js
tooltipRect.bottom - parentRect.bottom
```
use
```js
(tooltipRect.bottom - parentRect.bottom) / tooltipRect.height
```
Otherwise for example a short but wide tooltip might wind up on top or bottom even if it's mostly cut off and could be mostly visible on the left or right, where more pixels are cut off but that's a smaller fraction of its total area.